### PR TITLE
Pass position IDs when scoring online preference data

### LINKED ISSUE #6800: Online DPO / XPO / Nash-MD logprob forwards omit `position_ids` under left-padded prompts
### Observation

The logprob forwards in the Online DPO / XPO / Nash-MD trainers pass `input_ids` and `attention_mask` but no `position_ids`, while their prompts are left-padded. For models whose `transformers` implementation falls back to `torch.arange` when `position_ids is None`, this scores every padded row at positions shifted by its pad count.

`trl/experimental/online_dpo/online_dpo_trainer.py:1084`
```python
model_kwargs = {"attention_mask": prompt_completion_mask}
...
output = model(prompt_completion_ids, **model_kwargs)
```

`trl/experimental/xpo/xpo_trainer.py:275` and `trl/experimental/nash_md/nash_md_trainer.py:344`
```python
output = m(data["input_ids"], attention_mask=data["attention_mask"])
```

Prompts are left-padded at `online_dpo_trainer.py:694` and `:954` (`padding_side="left"`), and XPO/Nash-MD inherit that pipeline via `OnlineDPOTrainer`.

PPO is unaffected — its `forward()` builds them explicitly (`trl/experimental/ppo/ppo_trainer.py:212`):
```python
position_ids = attention_mask.cumsum(1) - attention_mask.long()
```
which appears to date to #469.

### Why it may be worth a note rather than a fix

For RoPE models this is a genuine no-op: relative offsets are preserved and pads are masked out. Since that covers essentially every model these trainers are used with in practice (and all of the test models), the omission is harmless today, which is presumably why it reads as the accepted convention.

The part that surprised me is that it is not simply a constant offset when it *does* bite. `generate()` derives positions from the mask (`generation/utils.py:721`, `position_ids = attention_mask.long().cumsum(-1) - 1`), so completions are sampled at mask-relative positions and then re-scored at `arange` positions. Generation and scoring disagree with each other, rather than both being consistently shifted.

### Repro

`transformers` 5.9.0, `torch` 2.12.0. Two prompts of different length, left-padded (13 pads on the short row), scoring the same completion with and without `position_ids=attention_mask.cumsum(1) - attention_mask.long()`, using the same slicing as `_forward`:

```
model=gpt2
  row 0  padded  (13 left pads)     max|delta logp| = 1.168e+01   sum|delta| = 3.400e+01
  row 1  unpadded (control)         max|delta logp| = 0.000e+00   sum|delta| = 0.000e+00
  sequence logprob (padded row): no-pos=-47.1921  with-pos=-13.1873  delta=-34.0049

model=Qwen/Qwen2-0.5B-Instruct
  row 0  padded  (13 left pads)     max|delta logp| = 3.815e-06   sum|delta| = 9.298e-06
  row 1  unpadded (control)         max|delta logp| = 0.000e+00   sum|delta| = 0.000e+00
  sequence logprob (padded row): no-pos=-11.0847  with-pos=-11.0847  delta=0.0000

model=facebook/opt-125m
  row 0  padded  (13 left pads)     max|delta logp| = 0.000e+00   sum|delta| = 0.000e+00
  row 1  unpadded (control)         max|delta logp| = 0.000e+00   sum|delta| = 0.000e+00
  sequence logprob (padded row): no-pos=-15.2500  with-pos=-15.2500  delta=0.0000
```

<details>
<summary>script</summary>

```python
import sys, torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id = sys.argv[1]
tok = AutoTokenizer.from_pretrained(model_id)
if tok.pad_token is None:
    tok.pad_token = tok.eos_token
model = AutoModelForCausalLM.from_pretrained(model_id, dtype=torch.float32).eval()

prompts = [
    "The capital of France is",
    "In a shocking turn of events yesterday afternoon, the committee announced that the capital of France is",
]
completion = " Paris, a city known for"

tok.padding_side = "left"
p = tok(prompts, return_tensors="pt", padding=True, add_special_tokens=False)
c_ids = tok([completion, completion], return_tensors="pt", add_special_tokens=False)["input_ids"]

input_ids = torch.cat([p["input_ids"], c_ids], dim=1)
mask = torch.cat([p["attention_mask"], torch.ones_like(c_ids)], dim=1)
plen = p["input_ids"].shape[1]

def completion_logprobs(**extra):
    with torch.no_grad():
        logits = model(input_ids=input_ids, attention_mask=mask, **extra).logits
    logits = logits[:, plen - 1 : -1]
    return torch.take_along_dim(logits.log_softmax(-1), c_ids.unsqueeze(-1), dim=2).squeeze(-1)

pos = mask.cumsum(1) - mask.long()
lp_trl, lp_fix = completion_logprobs(), completion_logprobs(position_ids=pos)

n_pad = (p["attention_mask"][0] == 0).sum().item()
print(f"model={model_id}")
for r, tag in [(0, f"padded  ({n_pad} left pads)"), (1, "unpadded (control)")]:
    d = (lp_trl[r] - lp_fix[r]).abs()
    print(f"  row {r} {tag:26s} max|delta logp| = {d.max():.3e}   sum|delta| = {d.sum():.3e}")
print(f"  sequence logprob (padded row): no-pos={lp_trl[0].sum():.4f}  "
      f"with-pos={lp_fix[0].sum():.4f}  delta={(lp_trl[0]-lp_fix[0]).sum():.4f}")
```
</details>

Worth flagging that the affected set is narrower than "models with absolute position embeddings". OPT has learned absolute positions but is unaffected, because `OPTLearnedPositionalEmbedding.forward` reconstructs them from the mask itself when none are passed (`models/opt/modeling_opt.py:64-68`) — the same formula PPO uses. GPT-2 has no such fallback (`models/gpt2/modeling_gpt2.py:579-582`), so it is the one that moves. The condition is really "the model's `transformers` implementation defaults to `arange`", which is a property of the modeling file rather than of the position-embedding scheme.

### Question

Is passing `position_ids` here something you'd want, or is relying on RoPE-only usage the intended contract? Happy to open a PR adding `position_ids = attention_mask.cumsum(1) - attention_mask.long()` to the three call sites to match PPO if it's wanted — it's cheap and a no-op for every currently tested model. Equally happy to leave it if the convention is deliberate.

Checked against `main` @ f724602; these files are unchanged since.

### PR DESCRIPTION
**Pull request by guptaishaan**
# What does this PR do?

Fixes #6800

Online DPO, XPO, and Nash-MD scored left-padded sequences without passing mask-relative `position_ids`. This made scoring positions disagree with generation positions for models such as GPT-2.

The affected forwards now derive `position_ids` from the attention mask. Regression tests cover all three trainers. Thanks to @mmjerge for reporting the issue and providing the GPT-2 repro.

Verified on Linux and CPU with an A40 host. The GPT-2 repro showed a 34.005 sequence log-prob difference before the fix. All 32 tests in the three trainer modules passed, with 13 existing skips, and all pre-commit hooks passed. GPU execution and the full repository suite were not run.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes how preference losses compute log-probs for padded sequences, which directly affects training gradients; scope is narrow and covered by new unit tests.
> 
> **Overview**
> Fixes incorrect log-probability scoring for **left-padded** batches in Online DPO, XPO, and Nash-MD (issue #6800). Scoring forwards previously omitted `position_ids`, so token positions could disagree with generation—especially on models like GPT-2.
> 
> **Online DPO** `_forward` now includes `position_ids` derived from the prompt+completion attention mask (`cumsum(mask) - mask`). **XPO** and **Nash-MD** apply the same rule in `_compute_logprobs` for policy and reference calls.
> 
> Regression tests assert mocked model forwards receive the expected `position_ids` for padded sequences.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 845b712319085740183989ce822e5c7eb89d66ae. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).
